tests/unittest_brain_dataclasses.py

```diff
@@ -991,6 +992,47 @@ class ChildWithMixedParents(BaseParent, NotADataclassParent):
     assert [a.name for a in mixed_init.args.args] == ["self", "_abc", "ghi"]
     assert [a.value for a in mixed_init.args.defaults] == [1, 3]
 
+    first, impossible = astroid.extract_node(
+        """
+    from dataclasses import dataclass
+
+    @dataclass
+    class BaseParent:
+        required: bool
+
+    @dataclass
+    class FirstChild(BaseParent):
+        ...
+
+    @dataclass
+    class SecondChild(BaseParent):
+        optional: bool = False
+
+    @dataclass
+    class GrandChild(FirstChild, SecondChild):
+        ...
+
+    @dataclass
+    class ThirdChild:
+        other: bool = False
+
+    @dataclass
+    class ImpossibleGrandChild(FirstChild, SecondChild, ThirdChild):
+        ...
+
+    GrandChild.__init__  #@
+    ImpossibleGrandChild() #@
+    """
+    )
+
+    first_init: bases.UnboundMethod = next(first.infer())
+    assert [a.name for a in first_init.args.args] == ["self", "required", "optional"]
+    assert [a.value for a in first_init.args.defaults] == [False]
+
+    # TODO: This should be an assertion, but the dataclass brain is a transform
+    # which currently can't return an Uninferable correctly.
+    assert not next(impossible.infer()) is Uninferable
```

Comment:
I'm having trouble following what "should be an assertion" means. You mean the assertion should have the opposite result? If so we should extract it into a test case and add an expected fail marker.

Reply:
Yeah I thought about extracting but this example kinda belongs with the other test. Can we `xfail` within a function? (That's basically what I'm doing with the `assert not` here).

Reply:
I'm not sure xfailing within a function helps, because you only get one result per (parameterized) test. That result should be xfailed. It's okay with me if you want to xfail the entire test.

Reply:
I can separate this test out, but doesn't that achieve the same as this `assert not`? The `# TODO` or `xfail` serves as a reminder to look into it and the comment or docstring explains that this is actually supposed to be the inverse of the current situation?

Reply:
I didn't find the `TODO` clear. And I think future developers will rely on the assertion (e.g. with git blame) to understand what the correct behavior was understood to be at the time. So I'm -1 on having an assert that asserts the incorrect behavior. Also, the xfail is useful because it creates an xpass when it's finally fixed, which can we can use as an alert to clean things up. (I think there are some xpasses currently we should do something about, as it happens...)